Documentation/Relationship/Relationship.md

```diff
@@ -0,0 +1,334 @@
+  "lastname": "Bauer",
+  "email": "mail@example.org",
+  "_links": {
+    "self": "https://example.api/user/472"
+  }
+}
+```
+
+_Note: to be fair the `self` link can also be more easily retrieved by looking at what the client sent in its request.
+However what we actually do is building a mapping from types to their respective `Handler`
+which is needed for the following chapters._
+
+### 3.2. `Handler`s which return an `Identifiable` but don't have a parameter
+
+A special case to [3.1](#31-looking-at-the-return-type-of-handle) is, when we detect that a `Handler` returns a type
+conforming to `Identifiable`, but doesn't have a `@PathParameter` which can be mapped to `.id` property.
+
+This case is illustrated by the following example web service with the `MeUserHandler`:
+```swift
+struct TestService: WebService, Component {
+  @PathParameter var userId: User.ID
+  @PathParameter var postId: User.ID
+
+  var content: some Component {
+    Group("user", $userId) {
+      UserHandler(userId: $userId)
+      // could contain other routes giving more information for the given user, e.g. ./posts
+    }
+    Group("me") {
+      MeUserHandler() // handle() returns `User`
+    }
+  }
+}
+```
+
+In such a case, we can search the web service for a `Handler` which also returns `User`, but has a `@PathParameter`.
+We could then find the `UserHandler` registered under `/user/:userId`.
+
+A response returned from the `MeUserHandler` would then include a `self` relationship pointing to `/user/{userId}` 
+(`userId` is known from the returned `User` instance). More specifically the response will include all relationships
+which are also returned on a request to `user/:userId` (meaning links to potential subroutes).
+
+**Open questions:**
+* This approach heavily relies on the fact that `/me` returns the same type as `user/:userId`. It might be worth
+  creating a explicit mechanism to define the `self` link in a case were `/me` would only return a subset of the properties
+  of `User`.  
+  This could be solved using the same mechanism as presented in [3.3](#33-defining-foreignkeys). 
+
+### 3.3. Defining `ForeignKeys`
+
+In this chapter we assume the existence of the example webservice from the previous chapter and adding a `/article`
+handler.
+
+```swift
+var content: some Component {
+  // ...
+  Group("article", $articleId) {
+    ArticleHandler(articleId: $articleId)
+  }
+}
+```
+
+The idea is, that every `Article` is written by a certain `User` (aka the author).
+A potential REST API response should then include a link to the respective `User` resource.  
+To be able to display such a relationship the user needs to provide that information to us.
+
+A `Article` could then be defined like the following:
+```swift
+struct Article: IdentifiableWithForeignKeys {
+  var id: String
+  var heading: String
+  var content: String
+
+  var writtenBy: String
+
+  static var foreignKeys: some ForeignKeys {
+    ForeignKey<User>(name: "author", at: \.writtenBy)
```

Comment:
I think this idea makes sense to indicate the links in the model. I would suggest to think about a Property wrapper approach. You could also have the same functionality if we have a `@ForeignKey` property wrapper around the existing property. This is the same approach that Fluent uses e.g. with `@Parent`: https://github.com/vapor/fluent-kit/blob/ef2a9a76c48856f0388ee2026d67d2ebde8c571c/Sources/FluentKit/Properties/Parent.swift and https://docs.vapor.codes/4.0/fluent/overview/ . I could image an approach that without a database imported we would have our own property wrapper that indicates the relationship and if @hendesi 's work is integrated and Fluent is imported the Fluent property wrappers will serve the exact same functionality and will be treated the same way as our own property wrappers. This would also make the move to a database a nice transition and the definition of `ForeignKey`s is not done twice in the model.

Reply:
I just remember the you mentioned some `Codable` problems with the property wrappers, maybe we can use the fluent approach to solve these problems. They also support `Codable` with their property wrappers.

Reply:
If I'm not wrong you are proposing the same thing I proposed to @nerdsupremacist on our meeting (for reference my first draft of my proposal https://gist.github.com/Supereg/0c46967264611f0712e82c25624b86f3)

This basically suggested a syntax like 
```swift
struct Article: Identifiable {
    var id: String
    @ForeignKey(for: User.self) var writtenBy: String // String is the same type as User.ID here
    var heading: String
    var content: String
}
```

> I just remember the you mentioned some Codable problems with the property wrappers, maybe we can use the fluent approach to solve these problems. They also support Codable with their property wrappers.

That was one of the reasons why we move to the sort of "Mini-DSL", which could also leave some room for future advancements.

Reply:
I really like the idea of having an `ForeignKey` or maybe a bit more descriptive as we are not in a database context without adding a database an `@Relationship` property wrapper that could be used like this: `@Relationship(to: User.self)`. I would argue that property wrappers would allow a easier transition to and combination with web services that use a database. If a developer would use an `@Parent` property wrapper from fluent this would already provide all the functionality needed and 

I see, that's why I remembered `Codable`, that's one topic that @nerdsupremacist and I talked about. I think that fluent solved this problem pretty elegantly using the Codable implementation they provide.